### ISSUE TITLE
writeSequence not well-tested

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10493]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#745|
|Type|defect|
|Created|2004-10-11 23:36:00Z|
|Branch|https://github.com/twisted/twisted/tree/test-write-sequence-745-7|

<details><summary>Searchable metadata</summary>

```
trac-id__745 745
type__defect defect
reporter__jknight jknight
priority__high high
milestone__ 
branch__branches_test_write_sequence_745_7 branches/test-write-sequence-745-7
branch_author__exarkun__therve exarkun, therve
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1097537760000000 1097537760000000
changetime__1317691293000000 1317691293000000
version__ 
owner__exarkun exarkun
cc__itamarst cc__jknight cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
r11925 broke writeSequence badly. The new writeSequence doesn't call startWriting, nor pause 
producers when too much is buffered. 

Since startWriting isn't called, if you just use writeSequence and not write, nothing will ever get written. 
Apparently our tests suck, because the tests don't catch this.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Fixed the behaviour, now it needs tests.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [23463]) Branching to 'test-write-sequence-745'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [23464]) Add tests for the 2 corner cases mentioned: no write and a streaming producer.

Refs [#10493](https://github.com/twisted/twisted/issues/10493)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @therve|
|-|-|

I've seen a problem with iocp I think.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

There was indeed a problem that I fixed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

I waited just long enough before reviewing this so I could ask if you felt like writing these new tests in `twisted/internet/test/test_tcp.py`. ;)

This might also be an okay time for me to mention a couple things I'm not sure about with `ReactorBuilder`.  One is that it doesn't make you do cleanup.  Maybe this is okay for direct reactor tests, since the tests should really know what they're doing, and leaving stuff behind afterwards can be okay as long as you know you're doing it.  The other is that `ReactorBuilder` tests don't support `Deferred` results as nicely as normal tests do.  If you return a `Deferred` from one, trial will spin the wrong reactor for you.  This isn't catastrophic, since you can start and stop your reactor inside one test method, but it does mean you don't get timeouts.  Anyway, that's probably enough off-topic rambling.

Nice IOCP fix.

`test_writeSequenceStreamingProducer` and `test_writeSequenceNonStreamingProducer` might be problematic.  It depends on the OS send-buffer size being less than 100000 bytes.  Generally it is (although Linux gets close, 87380) but it'd be best if we could avoid relying on this.  I'm not sure what I'd suggest doing instead.  A fake could pretend to have a full buffer, but faking socket is a big job (and we'd want a verified fake).  Maybe if the assumption being made is documented a bit more clearly (ie, that the socket send buffer is no bigger than 100000 bytes) it'll be fine.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [23652]) Branching to 'test-write-sequence-745-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

I've moved the tests. I wonder if the reactor manipulation is robust enough, but it works for now. I also solved a bug in tcp where global reactor was used.

Regarding the 100000 bytes in tests, they are calculated from the bufferSize of `FileDescriptor`, not from the OS limit. I've added a comment to explain that.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [25016]) Branching to 'test-write-sequence-745-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve removed owner|
|-|-|

I forgot why I didn't put this into review at that time, so I'll push it now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @therve|
|-|-|

  1. in `twisted/internet/tcp.py`, `Server` is changed so that `reactor` to its `__init__` is optional, and the caller is changed to pass a reactor.  Are both of these changes actually necessary?  I dislike the former and like the latter.
  1. I'm confused about why either `writeSequence` or `write` has any code for dealing with a producer at all.  Why isn't `doWrite` solely responsible for pausing and resuming a producer?  It's the method which knows if the socket is actually accepting more data or not.
  1. Assuming there's a good reason for that, or at least that it's something that should be fixed separately from this ticket, there's a ton of duplication here.  Four implementations of the producer pausing logic, only three of which are correct even with the fix in this branch:
    1. twisted.internet.abstract.FileDescriptor.write
    1. twisted.internet.abstract.FileDescriptor.writeSequence
    1. twisted.internet.iocpreactor.abstract.FileDescriptor.write
    1. twisted.internet.iocpreactor.abstract.FileDescriptor.writeSequence
  1. The tests should use `ReactorBuilder.runReactor`
  1. Since the testcase is named `WriteSequenceTests`, maybe the test methods don't need `writeSequence` in their names.
  1. There are failures on Windows (not even counting IOCP builders). :(  http://buildbot.twistedmatrix.com/builders/server-2k8-x86-py2.5-select/builds/12

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve removed owner|
|-|-|

OK, this is back to review. Builders look green enough. I think this ticket will trigger some tickets, but I'd like to defer this:
 * remove the duplication mentioned above
 * do something about `bufferSize` versus `writeBufferSize`

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

Cool.  Please file tickets for those two points, add a comment in the code referring to each of those tickets, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @therve|
|-|-|

Bump.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve commented|
|-|-|

The new tests I wrote are affected by the same bug as [#3925](https://github.com/twisted/twisted/issues/3925). So merging this would introduce 8 new test failures under Windows...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

Achtung.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

(In [28544]) Branching to 'test-write-sequence-745-4'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

Those tests are still broken under Windows...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [30270]) Branching to 'test-write-sequence-745-5'

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [32682]) Branching to 'test-write-sequence-745-6'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [32691]) Rewrite the new writeSequence/producer tests to avoid [[#5285](https://github.com/twisted/twisted/issues/5285)](https://github.com/twisted/twisted/issues/5285)

refs [#10493](https://github.com/twisted/twisted/issues/10493)
refs #5285

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

I merged this forward and resolved some simple conflicts.  I also changed `test_streamingProducer` and `test_nonStreamingProducer` to work - they were missing basic setup, like listenTCP and connectTCP, so I guess they were left in a slightly incomplete state last time this was worked on.  Then I changed them around a bit to avoid triggering [#5285](https://github.com/twisted/twisted/issues/5285).

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/test-write-sequence-745-6) look pretty good now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph set owner to @exarkun|
|-|-|

Looks like there's still one more trivial conflict, but otherwise this branch looks fabulous.  Love the IOCP fix; I wonder if it will make future builds more reliable?

Please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [32736]) Branching to 'test-write-sequence-745-7'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun set status to `closed`|
|-|-|

(In [32738]) Merge test-write-sequence-745-7

Author: therve, exarkun
Reviewer: exarkun, glyph
Fixes: [#10493](https://github.com/twisted/twisted/issues/10493)

Add `ReactorBuilder`-style unit tests for `ITransport.writeSequence`.  Fix a bug
thusly revealed in IOCP reactor's implementation of that method when combined
with a pull producer.
